File: Assignment 4/Solutions/Huffman Tree.java
import java.util.*;
import java.util.Map;

class HuffmanNode implements Comparable<HuffmanNode> {
    int freq;
    Character character;
    HuffmanNode left;
    HuffmanNode right;

    public HuffmanNode(int freq, Character character, HuffmanNode left, HuffmanNode right) {
        this.freq = freq;
        this.character = character;
        this.left = left;
        this.right = right;
    }

    @Override
    public int compareTo(HuffmanNode other) {
        return this.freq - other.freq;
    }
}

public class Main {
    private static Map<Character, String> codeDict = new HashMap<>();

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();

        // Frequency for distinct chars
        Map<Character, Integer> freqDict = new HashMap<>();
        for (char c : input.toCharArray()) {
            freqDict.put(c, freqDict.getOrDefault(c, 0) + 1);
        }

        if (freqDict.size() == 1) {
            System.out.println(input.length());
            return;
        }

        // Construct Huffman tree
        PriorityQueue<HuffmanNode> freqHeap = new PriorityQueue<>();
        for (Map.Entry<Character, Integer> entry : freqDict.entrySet()) {
            freqHeap.add(new HuffmanNode(entry.getValue(), entry.getKey(), null, null));
        }

        while (freqHeap.size() != 1) {
            HuffmanNode left = freqHeap.poll();
            HuffmanNode right = freqHeap.poll();
            HuffmanNode node = new HuffmanNode(left.freq + right.freq, null, left, right);
            freqHeap.add(node);
        }

        // Construct codes
        HuffmanNode root = freqHeap.peek();
        createCode(root, "");

        // Calculate total bits
        int bits = 0;
        for (char c : input.toCharArray()) {
            bits += codeDict.get(c).length();
        }
        System.out.println(bits);
    }

    private static void createCode(HuffmanNode start, String curCode) {
        if (start != null && start.character != null) {
            codeDict.put(start.character, curCode);
        } else if (start != null) {
            createCode(start.left, curCode + "0");
            createCode(start.right, curCode + "1");
        }
    }
}

File: Assignment 4/Solutions/Huffman Tree.py
import heapq

class HuffmanNode:
    def __init__(self, freq, char = None, left = None, right = None):
        self.freq = freq
        self.char = char
        self.left = left
        self.right = right
    
    def __lt__(self, other):
        return self.freq < other.freq
    
string = input()

# Frequency for distinct chars.
freq_dict = {}
for char in string:
    freq_dict[char] = freq_dict.get(char, 0) + 1

if len(freq_dict) == 1:
    print(len(string))
else:
    # Construct Huffman tree.
    freq_heap = [HuffmanNode(freq, char) for char, freq in freq_dict.items()]
    heapq.heapify(freq_heap)

    while len(freq_heap) != 1:
        l, r = heapq.heappop(freq_heap), heapq.heappop(freq_heap)
        node = HuffmanNode(l.freq + r.freq, left = l, right = r)
        heapq.heappush(freq_heap, node)
    
    def create_code(start, cur_code):
        if (start is not None) and (start.char is not None):
            code_dict[start.char] = cur_code
        elif start is not None:
            create_code(start.left, cur_code + '0')
            create_code(start.right, cur_code + '1')

    # Construct codes.
    root = freq_heap[0]
    code_dict = {}
    create_code(root, '')

    bits = sum(len(code_dict[char]) for char in string)
    print(bits)
